sim/src/main/scala/spinal/sim/VerilatorBackend.scala

```diff
@@ -548,10 +615,27 @@ JNIEXPORT void API JNICALL ${jniPrefix}disableWave_1${uniqueId}
     }
   }
 
+  def copyOutputToCache(): Unit = {
```

Comment:
copyOutputToCache should be executed atomicaly (global mutex), in case of multiple compilation update the cache at the same time ?

Reply:
Dou you mean the cache could be updated at the same time when running the compiled scala program two times concurrently?

In case you mean to just run `simCfg.compile()` more than one time, the passed directory `cacheDir` is also renamed. So in this case there should be no conflicts.

Reply:
> Dou you mean the cache could be updated at the same time when running the compiled scala program two times concurrently?

Yes, right, basicaly, a mutex defined in a scala object could be locked / unlocked when checking / copying / write the cache.

Reply:
I don't understand why that is needed. When running the program multiple times concurrently they would run verilator in the same directory (please correct me if I'm wrong). So this would fail anyway.

Can you give an example which shows where the current implementation (of the cache) would fail but should succeed?

Reply:
> When running the program multiple times concurrently they would run verilator in the same director

It won't, basicaly there is a protection against that, and if a natural workspace name is Foo, then the second time that workspace name is asked, it will get Foo_1 then Foo_2.

Basicaly, that was put in place to allow aggressive multithreading when running regressions in parallel.

> Can you give an example which shows where the current implementation (of the cache) would fail but should succeed?

So my guess is that so far, two compilations of the same test could consider that the cache is empty at the same time (which is fine for me):

https://github.com/SpinalHDL/SpinalHDL/blob/6f3763730c20379395ef935b49e9e76a32631e27/sim/src/main/scala/spinal/sim/VerilatorBackend.scala#L526

The issue would come later, when both will try to update the same cache entry at the same time ?

So to avoid that, just having a mutex to protect the cache read / copy / write operations should be enough i guess.

Reply:
> > When running the program multiple times concurrently they would run verilator in the same director
> 
> It won't, basicaly there is a protection against that, and if a natural workspace name is Foo, then the second time that workspace name is asked, it will get Foo_1 then Foo_2.

That protection should also apply to the current implementation of the cache. At least when calling `.compile()` multiple times there will also be created separate directories for `Foo_1`, `Foo_2`, ... in the cache. So two threads will never try to write/read to/from the same files.

But now I realize that this behaviour is not wanted because verilator compilation is done `n` times in the first run (when running `n` parallel test). To fix this I have to use a single cache (directory) for a module and then I can / have to protect the cache using a mutex. Right?

Just for clarification: This only applies when running mulitple tests in a single process? Or does the protection (which renames the workspaces) also work when running multiple tests but each in its own process?

Reply:
Done, hope I understood you correctly

Reply:
Ahhhhh, i were assuming the cache was global, that was the missunderstanding

> To fix this I have to use a single cache (directory) for a module and then I can / have to protect the cache using a mutex. Right?

I would say instead, we could have a single folder containing all caches. Each entry in that cache would have the hash as folder name. That way, nothing is related to specific modules. The key would be enough to ensure we pick the proper cache entry. This would easily support having the same module with different parameters being cached.

> Just for clarification: This only applies when running mulitple tests in a single process? Or does the protection (which renames the workspaces) also work when running multiple tests but each in its own process?

This only applies when running mulitple tests in a single process. If you run two JVM, this won't work.

Reply:
> I would say instead, we could have a single folder containing all caches. Each entry in that cache would have the hash as folder name. That way, nothing is related to specific modules. The key would be enough to ensure we pick the proper cache entry. This would easily support having the same module with different parameters being cached.

Ok, but then we need a way to automatically delete unused cache entries. Otherwise the cache can get really large in a short time. Maybe specify a maximum number of cache entries and delete the directory with the oldest modification time when a cache entry has to be replaced. Or any other idea?

Reply:
Right, To me we can keep about 100 in cache, and when there is more, it can either randomly delete one, either pick the one with the oldest date as you proposed ^^
As you want :)

Reply:
done